pkg/util/scrubber/yaml_scrubber.go

```diff
@@ -0,0 +1,89 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package scrubber
+
+import (
+	"fmt"
+	"os"
+
+	"gopkg.in/yaml.v2"
+)
+
+type scrubCallback = func(string, interface{}) (bool, interface{})
+
+func walkSlice(data []interface{}, callback scrubCallback) {
+	for _, k := range data {
+		switch v := k.(type) {
+		case map[interface{}]interface{}:
+			walkHash(v, callback)
+		case []interface{}:
+			walkSlice(v, callback)
+		}
+	}
+}
+
+func walkHash(data map[interface{}]interface{}, callback scrubCallback) {
+	for k, v := range data {
+		if keyString, ok := k.(string); ok {
+			if match, newValue := callback(keyString, v); match {
+				data[keyString] = newValue
+				continue
+			}
+		}
+
+		switch v := data[k].(type) {
+		case map[interface{}]interface{}:
+			walkHash(v, callback)
+		case []interface{}:
+			walkSlice(v, callback)
+		}
+	}
+}
+
+// walk will go through loaded yaml and call callback on every strings allowing
+// the callback to overwrite the string value
+func walk(data *interface{}, callback scrubCallback) {
+	switch v := (*data).(type) {
+	case map[interface{}]interface{}:
+		walkHash(v, callback)
+	case []interface{}:
+		walkSlice(v, callback)
+	}
+}
+
+// ScrubYaml scrubs credentials from the given YAML by loading the data and scrubbing the object instead of the
+// serialized string.
+func (c *Scrubber) ScrubYaml(input []byte) ([]byte, error) {
+	var data *interface{}
+	err := yaml.Unmarshal(input, &data)
+
+	// if we can't load the yaml run the default scrubber on the input
+	if err == nil {
```

Comment:
As an interface, I feel it would make more sense to return an error here and let the caller decide how it should handle the error (ie. call the default scrubber). I'm not sure, if you feel it'd make the code so much more verbose this is fine.

Reply:
I agree it's better but we want to still apply the text scrubber with or without the YAML scrubber succeeding, We could let the caller call both but I feel it's a error-prone.